cli/crates/federated-dev/src/dev/router.rs

```diff
@@ -1,28 +1,84 @@
-use super::bus::GraphReceiver;
+use std::sync::Arc;
+
+use super::bus::{GraphReceiver, RequestReceiver, ResponseSender};
+use engine::ServerError;
+use engine_v2::Engine;
+use futures_concurrency::stream::Merge;
+use futures_util::{stream::BoxStream, StreamExt};
+use graphql_composition::FederatedGraph;
+use tokio_stream::wrappers::ReceiverStream;
 
 pub(crate) struct Router {
-    bus: GraphReceiver,
+    graph_bus: GraphReceiver,
+    request_bus: RequestReceiver,
+    engine: Option<Arc<Engine>>,
 }
 
-/// TODO: Benjamin and Graeme!
-///
-/// Add the router here, route the queries from the Axum to it.
-/// When the handler receives a new graph, restart the router.
-///
-/// P.s. there might be a need for a m-m-mutex...
 impl Router {
-    pub fn new(bus: GraphReceiver) -> Self {
-        Self { bus }
+    pub fn new(graph_bus: GraphReceiver, request_bus: RequestReceiver) -> Self {
+        Self {
+            graph_bus,
+            request_bus,
+            engine: None,
+        }
     }
 
     pub async fn handler(mut self) -> Result<(), crate::Error> {
-        while let Some(graph) = self.bus.recv().await {
-            // remvove this when adding the actual router
-            let rendered = graphql_composition::render_sdl(&graph).expect("must render");
-            println!("******** GRAPH ********");
-            print!("{rendered}");
+        log::trace!("starting the router handler");
+
+        let streams: [RouterStream; 2] = [
+            Box::pin(ReceiverStream::new(self.graph_bus).map(RouterMessage::Graph)),
+            Box::pin(ReceiverStream::new(self.request_bus).map(RouterMessage::request)),
+        ];
+
+        let mut stream = streams.merge();
+
+        while let Some(message) = stream.next().await {
+            match (message, self.engine.as_ref()) {
+                (RouterMessage::Graph(graph), _) => {
+                    log::trace!("router got a new graph");
+
+                    self.engine = Some(new_engine(graph));
+                }
+                (RouterMessage::Request(request, response_sender), Some(engine)) => {
+                    log::trace!("router got a new request with an existing engine");
+
+                    tokio::spawn(run_request(request, response_sender, Arc::clone(engine)));
+                }
+                (RouterMessage::Request(_, response_sender), None) => {
+                    log::trace!("router got a new request with a missingengine");
```

Comment:
nitpick: missing engine

Reply:
don't want to wait on ci again so i'll fix this in next pr